### DISCUSSION TITLE
[Feature] Switcher to allow choosing "internal" or "external" endpoint URLs in mobile app

### DISCUSSION THREAD

**Discussion by jon-f-novabackup** [immich-app/immich#2394] [Feature Request]
### The feature

[Feature] Switcher to allow choosing "internal" or "external" endpoint URLs in mobile app (allow configuring a setting for the internal and external endpoints in the app). Right now to change endpoint URLs, like for use internally versus externally you must logout of the mobile app, and then you have to manually input a potentially hard to remember or lengthy endpoint address, likely with different ports, and then when you need to go back to internal you have to do that all over again. Both the internal and external URL would be a configurable setting in the mobile app, so somewhere in the app itself at least the endpoint URLs can be custom defined. If both endpoint addresses are configured then on app load each time it would prompt the user which endpoint to connect to, and both of the endpoint URL addresses would be listed next to [Internal]: and [External]:. The user would then click on either internal or external in the switcher list to attempt to connect to that endpoint URL. If this setting is not defined at all in the app, to custom define the "internal" and "external" endpoint URLs, then the traditional method of manually inputting the endpoint URL on the initial dialog would be used and that stays the same. If the user has custom defined the two endpoint URLs then the switcher would be presented each time the app is started. That would be the simplest method of implementation that I can think of for this feature. There is more discussion, for further context and explanation, about this feature request in the "general-discusssion" Discord chat from 05/07/23 at 10:21 AM PDT, and then again the same day later resuming at 3:12 PM PDT; the 'BVD_WhiteyTighties' and 'Vodros' users had some good input in that group discussion.

To expand further on this feature request, there could be an added preference checkbox, in mobile app settings, that sets the preferred order for an automatic connection to the endpoints, for both internal or external endpoint URLs, as the endpoint connection to always attempt first on app start, either the internal first and then the external, or vice-versa, and then if the timeout (which could just use a generic 30 seconds or so hardcoded value) is reached it then attempts the other endpoint URL to connect. There could also be a configurable timeout setting in the same mobile app settings UI, to custom set the value in a particular range of seconds for instance.

This feature could obviously be expanded on further in the future, but it would be a good start just to implement the switcher to be able to via the mobile app, depending on if you are at home or remote, be able to select either the "internal' or "external" endpoint URL at mobile app start. There are obviously other ways of trying to implement this as a homelab server admin but the difficulty in setup that could work for everyone outweighs how the Immich mobile app could be made to work to apply to everyone out of the box, with super simple configuration as a single setting in mobile app UI, to specify the two endpoint URLs if that use case is even needed for that Immich server instance, and avoids maintaining your own DNS servers, including setup of split DNS, that would need to be configured on every mobile client device (including to be able to also work for family members outside of your home router), etc.

### Platform

- [ ] Server
- [ ] Web
- [X] Mobile

**Comment by bo0tzz** [comment]
Duplicate of #308

**Reply by jon-f-novabackup** [reply]
I think that there is a reason why this request keeps coming up. There is great difficulty in setting up a split DNS, not everyone that can load a docker container in Linux will be able to setup their network to handle this, and you aren't even speaking about those that have family members that also login from home and remote and only use the Immich mobile app, and even if you do somehow host your own DNS servers (for split DNS) then your friends and family members that may not even live in the same location or country, and hence don't use your router, will need to utilize your DNS servers across all of their mobile devices, or utilize the same tunnel as you do always. This is a lot for a small self-hosting admin to manage versus simply having an internal endpoint URL and an external endpoint URL setting in the mobile app to only have to configure one time per device mobile app, to be able to handle situations where you want to utilize the local IP address method (no tunnel or VPN) versus external tunneled or otherwise DNS address method.

It seems like the response from the devs when this feature is asked about (it has been brought up at least 3 prior times I do see now) is to just setup your server to be able to handle it directly in DNS (like to setup split DNS), to better optimize the connection speed if using the app internally or externally, but in actuality it would be super simple just to have a way for the user to specify one endpoint URL for internal use and another endpoint URL for the external use, as a setting in the mobile app, so that the traffic and connection could be optimized depending on the mobile user's current physical location. Not all users of a single Immich instance are living in the same home in the same router to facilitate this easily otherwise. It also makes testing hosting with different methods very tedious if the mobile app doesn't have an easy way of switching URLs depending on where you are at or what you want to do currently to switch the endpoint URL quickly, unlike web browser bookmarks that could have the address already predefined per connection method and named as such and easy to pick one. I think it is important to realize that not everyone that can host an Immich server are networking experts. This switch endpoint capability in the mobile app speaks to that need. I think that the amount of time needed to implement this in the mobile app at the most basic level outweighs the hundreds of hours of time it would take the 1% of the user base to hunt through networking setup guides to get split DNS or otherwise working perfectly, and after which that special setup has to be maintained across all client mobile devices (think friends and family as well), as compared to having a simple internal endpoint to external endpoint switcher in the mobile app to choose from, if the user needs it (by configuring the two endpoints URLs in settings first). I don't get the reluctance to directly address this feature to implement a simple URL switcher, and it seems that all of the prior requests for this feature were closed as not needed.
